arch/arm/core/thread.c

```diff
@@ -67,31 +78,68 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 	pInitCtx = (struct __esf *)(STACK_ROUND_DOWN(stackEnd -
 						     sizeof(struct __esf)));
 
-	pInitCtx->pc = ((u32_t)_thread_entry) & 0xfffffffe;
+#if CONFIG_ARM_USERSPACE
+	if (options & K_USER) {
+		pInitCtx->pc = (u32_t)_arch_user_mode_enter;
+	} else {
+		pInitCtx->pc = (u32_t)_thread_entry;
+	}
+#else
+	pInitCtx->pc = (u32_t)_thread_entry;
+#endif
+
+	/* force ARM mode by clearing LSB of address */
+	pInitCtx->pc &= 0xfffffffe;
+
 	pInitCtx->a1 = (u32_t)pEntry;
 	pInitCtx->a2 = (u32_t)parameter1;
 	pInitCtx->a3 = (u32_t)parameter2;
 	pInitCtx->a4 = (u32_t)parameter3;
 	pInitCtx->xpsr =
 		0x01000000UL; /* clear all, thumb bit is 1, even if RO */
 
-#ifdef CONFIG_THREAD_MONITOR
-	/*
-	 * In debug mode thread->entry give direct access to the thread entry
-	 * and the corresponding parameters.
-	 */
-	thread->entry = (struct __thread_entry *)(pInitCtx);
-#endif
-
 	thread->callee_saved.psp = (u32_t)pInitCtx;
 	thread->arch.basepri = 0;
 
+#if CONFIG_ARM_USERSPACE
+	thread->arch.mode = 0;
+	if (options & K_USER) {
```

Comment:
Can't set priv_stack_start and priv_stack_size as zero here here even it is a privileged thread, since a privileged thread could drop it privilege later by calling k_thread_user_mode_enter().

Reply:
Maybe we could levave the code here as it is but just add extra code in _arch_user_mode_enter() to
set priv_stack_start and priv_stack_size when a privileged thread wants to drop its privilege.

Reply:
that's fine.  is that even a use case though running as privileged for a while then dropping?  I thought this was all decided at the point the thread is created

Reply:
I think sometimes we may create threads as privileged and drop their privilege after setting up memory domain
or access right of kobjects.

Reply:
> e. is that even a use case though running as privileged for a while then dropping

Yes it definitely is.
The API is k_thread_user_mode_enter()